condor_git_config.py

```diff
@@ -91,127 +98,174 @@ class ConfigCache(object):
+            dir_names.sort()
+            dir_path = Path(dir_path)
             for file_name in sorted(file_names):
-                rel_path = os.path.normpath(os.path.relpath(os.path.join(dir_path, file_name), repo_path))
+                rel_path = (dir_path / file_name).relative_to(repo_path)
                 if rel_path in seen:
                     continue
                 seen.add(rel_path)
                 yield rel_path
 
-    @property
-    def outdated(self):
-        try:
-            with open(self._meta_file, 'r') as raw_meta:
-                meta_data = json.load(raw_meta)
-        except FileNotFoundError:
-            return True
-        else:
-            if meta_data['git_uri'] != self.git_uri or meta_data['branch'] != self.branch:
-                LOGGER.critical('cache %r corrupted by other hook: %r', self, meta_data)
-                raise RuntimeError('config cache %r used for conflicting hooks' % self.cache_path)
-            else:
-                return meta_data['timestamp'] + self.max_age <= time.time()
-
-    def _update_metadata(self):
-        with open(self._meta_file, 'w') as raw_meta:
-            json.dump({'git_uri': self.git_uri, 'branch': self.branch, 'timestamp': time.time()}, raw_meta)
-
-    def refresh(self):
-        if not self.outdated:
+    def _refresh(self):
+        assert self._config_meta is not None
+        self._config_meta["reads"] += 1
+        if self._config_meta["timestamp"] + self.max_age > time.time():
             return
         repo_path = self.repo_path()
-        if not os.path.exists(os.path.join(repo_path, '.git')):
+        self._config_meta["pulls"] += 1
+        if not self.repo_path(".git").exists():
+            branch = [] if not self.branch else ["--branch", self.branch]
             subprocess.check_output(
-                ['git', 'clone', '--quiet', '--branch', str(self.branch), str(self.git_uri), repo_path],
+                ["git", "clone", "--quiet", *branch, self.git_uri, str(repo_path)],
                 timeout=30,
                 universal_newlines=True,
             )
+            self._config_meta["timestamp"] = time.time()
         else:
-            subprocess.check_output(
-                ['git', 'pull'],
-                timeout=30,
-                cwd=repo_path,
-                universal_newlines=True
-            )
-        self._update_metadata()
+            try:
+                subprocess.check_output(
+                    ["git", "pull"], timeout=30, cwd=repo_path, universal_newlines=True
+                )
+            except (subprocess.CalledProcessError, subprocess.TimeoutExpired):
+                pass
+            else:
+                self._config_meta["timestamp"] = time.time()
 
 
 class ConfigSelector(object):
-    def __init__(self, pattern, blacklist, whitelist, recurse):
+    """
+    Selector for a configuration file iterator
+    """
+
+    def __init__(self, pattern, blacklist, whitelist, recurse: bool):
```

Comment:
Please consider adding the correct type hints.

Reply:
Done.